### PR TITLE
Transactions: Request for initial feedback

### PR DESCRIPTION
**Pull request by asqui**

### CODE REVIEW

File: psycopg3/transaction.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+from typing import Optional, Type
+
+from . import errors as e
+from .pq import TransactionStatus
+
+
+class Transaction:
+    def __init__(self, conn: "Connection") -> None:
+        self._conn = conn
+        # self._original_autocommit: bool
```
Comment (asqui) [inline psycopg3/transaction.py]: TODO: mypy

File: psycopg3/connection.py [hunk 2]
```diff
@@ -110,6 +122,10 @@ def autocommit(self, value: bool) -> None:
     def _set_autocommit(self, value: bool) -> None:
         # Base implementation, not thread safe
         # subclasses must call it holding a lock
+        if len(self._savepoints) > 0:
+            raise e.ProgrammingError(
+                "can't change autocommit state when in Transaction context"
```
Comment (asqui) [inline psycopg3/connection.py]: How does this message sound?
```suggestion
                "couldn't change autocommit state: connection.transaction() context in progress"
```

Intended to better match wording of the error message for similar case, below:
```"couldn't change autocommit state: connection in transaction status X"```

And to more concretely relate the error message to the API invoked by user.

File: psycopg3/transaction.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
+"""
+Transaction context managers returned by Connection.transaction()
+"""
+
+# Copyright (C) 2020 The Psycopg Team
+
+# TODO: Should this module be _underscored since it's implementation detail?
```
Comment (asqui) [inline psycopg3/transaction.py:7]: Suggest moving `Transaction` class into `connection.py` because:
1. It's an implementation detail of `Connection`.
2. All the tests are in `test_connection.py` since that's where the functionality is exposed to the user.
3. ~~It prevents circular import complications with type annotations.~~
There's always going to be a chicken-and-egg problem requiring `Transaction` to use forward-references to `Connection` for type annotations.

File: psycopg3/connection.py [hunk 4]
```diff
@@ -85,6 +94,9 @@ def __init__(self, pgconn: pq.proto.PGconn):
         self._notify_handlers: List[NotifyHandler] = []
         # name of the postgres encoding (in bytes)
         self._pgenc = b""
+        # stack of savepoint names managed by active Transaction() blocks
+        # (The first item may be None, representing an outer transaction)
+        self._savepoints: List[Union[None, str]] = []
```
Comment (asqui) [inline psycopg3/connection.py]: I'm not thrilled with this being a list of savepoint names or, optionally, for the first item only, a marker (currently `None`) indicating top-level transaction. 

However I can't think of a better way to give Connection a way to know if any Transaction contexts are active (so that it can deny access to commit/rollback/autocommit) -- currently the criteria is if `self._savepoints` is non-empty, then transaction contexts are active.

It's worth noting that entering a transaction context may:
1. Begin an outer transaction (if one isn't already in progress)
2. Begin an outer transaction _and_ create a savepoint (if one is named -- though names aren't implemented yet)
3. Create a savepoint (if a transaction is already in progress)

So there isn't even a 1-to-1 correspondence between items in this list and Transaction instances, because in case 2 a single Transaction instance may result in two items in this list.

(This complexity is part of the reason that we decided to have psycopg-nestedtransactions _always_ create a savepoint, because it then takes away case 1, and therefore you don't need to have a None marker-item in the list because you'll get a savepoint there to make it non-empty. BUT, we've said that the overhead of creating a savepoint every time you begin a transaction is unnecessary overhead that is not appropriate, and so we find ourselves here...)

File: psycopg3/connection.py (hunk 2 above)
Comment (dvarrazzo) [reply psycopg3/connection.py]: Seems good :+1: Postgres has a clear distinction between "can not" (something impossible to obtain) and "could not" (something that it that moment wasn't possible to do) and I'd like to follow the same guideline.

File: psycopg3/transaction.py (hunk 3 above)
Comment (dvarrazzo) [reply psycopg3/transaction.py:7]: With type annotation there are plenty of circular imports. I guess you wrote this before discovering `if TYPE_CHECKING`?

Comment (dvarrazzo) [reply psycopg3/transaction.py:7]: I am ok to move the tests into a module of its own. At the moment there is no `_module.py`, and other objects are details (transform, waiting...) Let's leave it this way for now, then we can make a rule about things.

File: psycopg3/connection.py (hunk 4 above)
Comment (dvarrazzo) [reply psycopg3/connection.py]: It's only a placeholder. You could use an empty string or a string symbolic constant. Or have an `Optional[List[str]]` and give different meaning to None and empty list. I also feel that maybe we should use `pgconn.transaction_status` to detect if we are inside or outside the first level transaction.

File: psycopg3/connection.py [hunk 5]
```diff
@@ -8,7 +8,14 @@
 import logging
 import asyncio
 import threading
-from typing import Any, AsyncGenerator, Callable, Generator, List, NamedTuple
+from typing import (
```
Comment (dvarrazzo) [inline psycopg3/connection.py]: Why has this changed? BTW, I hate these vertical lists in import: I prefer to have more than one import statement from the same module rather than it.

File: psycopg3/connection.py [hunk 6]
```diff
@@ -110,6 +121,10 @@ def autocommit(self, value: bool) -> None:
     def _set_autocommit(self, value: bool) -> None:
         # Base implementation, not thread safe
         # subclasses must call it holding a lock
+        if len(self._savepoints) > 0:
```
Comment (dvarrazzo) [inline psycopg3/connection.py]: `if self._savepoints` is more idiomatic to check non-empty list. But I think this check should be really against `pgconn.transaction_status`.

File: psycopg3/connection.py [hunk 7]
```diff
@@ -256,33 +271,44 @@ def _start_query(self) -> None:
...
             return
 
-        self.pgconn.send_query(b"begin")
-        (pgres,) = self.wait(execute(self.pgconn))
-        if pgres.status != ExecStatus.COMMAND_OK:
-            raise e.OperationalError(
-                "error on begin:"
-                f" {pq.error_message(pgres, encoding=self.codec.name)}"
-            )
+        self._exec(b"begin")
```
Comment (dvarrazzo) [inline psycopg3/connection.py]: Let's call this `_exec_command` as it should be only used for commands rather than queries

File: psycopg3/connection.py [hunk 8]
```diff
@@ -256,33 +271,44 @@ def _start_query(self) -> None:
...
 
     def commit(self) -> None:
-        self._exec_commit_rollback(b"commit")
+        with self.lock:
+            if len(self._savepoints) > 0:
```
Comment (dvarrazzo) [inline psycopg3/connection.py]: this time we really care about the difference whether a transaction was started by a context or by the automatic dbapi begin, so we have to use `self._savepoints` rather than `pgconn.transaction_status`. Please use `if self._savepoints` or if you think `Optional[List[str]]` is a better idea it would be `self._savepoints is not None`

File: psycopg3/transaction.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+from .pq import TransactionStatus
+
+if TYPE_CHECKING:
+    from .connection import Connection
+
+
+class Transaction:
+    def __init__(self, conn: "Connection") -> None:
+        self._conn = conn
+        self._outer_transaction = False
```
Comment (dvarrazzo) [inline psycopg3/transaction.py:20]: Maybe you can do without `_outer_transaction` because that's equivalent to having `_savepoint_name is None`?

File: psycopg3/transaction.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+        with self._conn.lock:
+            if self._conn.pgconn.transaction_status == TransactionStatus.IDLE:
+                self._outer_transaction = True
+                self._savepoint_name = None
+
+                self._original_autocommit = self._conn.autocommit
+                if self._conn._autocommit:
+                    self._conn._autocommit = False
+
+                self._conn._start_query()
```
Comment (dvarrazzo) [inline psycopg3/transaction.py]: Uhm... are you doing the `_original_autocommit` here just because `_start_query()` would otherwise bail out? That `conn._start_query()` might actually be called `_maybe_begin()`, and here you are doing a job partially overlapping, by checking the `transaction_status`.

As on exit you are executing your COMMIT or ROLLBACK, you may as well execute a BEGIN here, instead of calling `_start_query()`.

File: psycopg3/transaction.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+        return self._conn
+
+    def __enter__(self) -> None:
+        with self._conn.lock:
+            if self._conn.pgconn.transaction_status == TransactionStatus.IDLE:
+                self._outer_transaction = True
+                self._savepoint_name = None
+
+                self._original_autocommit = self._conn.autocommit
+                if self._conn._autocommit:
```
Comment (dvarrazzo) [inline psycopg3/transaction.py]: So, because of what pointed out below, I don't think you need to change this, and I don't think that changing `_autocommit` is desirable. However, if it was, you don't need this if.

File: psycopg3/transaction.py [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+                    self._conn._autocommit = False
+
+                self._conn._start_query()
+                assert len(self._conn._savepoints) == 0
+                self._conn._savepoints.append(None)
+            else:
+                self._outer_transaction = False
+                self._savepoint_name = "tx_savepoint_{}".format(
+                    len([s for s in self._conn._savepoints if s is not None])
```
Comment (dvarrazzo) [inline psycopg3/transaction.py]: This makes a O(1) operation into an O(n) operation. Why bothering counting the exactly non-null savepoints? You only need them unique (and we verified actually not even that is required... but let's say we want them unique).

How about `self._savepoint_name = f"tx_savepoint_{len(self._conn_savepoints)}"`?

File: psycopg3/transaction.py [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+                if self._savepoint_name:
+                    self._conn._savepoints.pop()
+                    # TODO: Add test for this assert
+                    # assert self._conn._savepoints.pop() == self._savepoint
+                    self.__exec(f"release savepoint {self._savepoint_name}")
+                if self._outer_transaction:
+                    self._conn._savepoints.pop()
+                    # TODO: Add test for this assert
+                    # assert self._conn._savepoints.pop() is None
+                    self.__exec("commit")
```
Comment (dvarrazzo) [inline psycopg3/transaction.py]: I don't like the mix of bytes and strings in these `Connection._exec` and `Transaction.__exec`. I would rather have the former take both, with an `Union[bytes, str]` signature and an `if isinstance(command, str): command = command.encode("ascii")` and drop the latter, otherwise leave only `bytes` in the signatures and deal with everythin at bytes level (making `_savepoint_name` bytes and avoiding f-strings)

File: tests/test_connection.py [hunk 14]
```diff
@@ -8,6 +11,21 @@
 from psycopg3.conninfo import conninfo_to_dict
 
 
+@pytest.fixture
+def temp_table(svcconn):
```
Comment (dvarrazzo) [inline tests/test_connection.py]: This is too generic for such a specific `id text primary key` result.

Probably this fixture should return an object to create temp tables, so that tests would be:
```python
def test(temp_table):
    temp_table.create("foo", "id int primary key")
    # ...
```

File: tests/test_connection.py [hunk 15]
```diff
@@ -341,3 +349,274 @@ def cb1(n):
 
     with pytest.raises(ValueError):
         conn.remove_notify_handler(cb1)
+
```
Comment (dvarrazzo) [inline tests/test_connection.py]: Even if these tests are on the connection objects, I think they should be in a test module of its own. A 622 lines module, destined to increase as there are several features missing, I'd say it's too much.

File: tests/test_connection.py [hunk 16]
```diff
@@ -341,3 +349,274 @@ def cb1(n):
...
+def test_transaction(conn):
+    """Basic use of transaction() to BEGIN and COMMIT a transaction."""
+    assert conn.pgconn.transaction_status == conn.TransactionStatus.IDLE
+    with conn.transaction():
+        assert conn.pgconn.transaction_status == conn.TransactionStatus.INTRANS
+    assert conn.pgconn.transaction_status == conn.TransactionStatus.IDLE
+
+
+def test_transaction_exposes_associated_connection(conn):
+    """Transaction exposes it's connection as a read-only property."""
```
Comment (dvarrazzo) [inline tests/test_connection.py]: "its" connection

File: tests/test_connection.py [hunk 17]
```diff
@@ -341,3 +349,274 @@ def cb1(n):
...
+    tx = conn.transaction()
+    assert tx.connection is conn
+    with pytest.raises(AttributeError):
+        tx.connection = conn
+
+
+def test_transaction_is_not_kept_alive_by_connection(conn):
+    """
+    Since the Transaction has a reference back to the connection, ensure that
+    the opposite is not also true, as we want to avoid reference cycles.
```
Comment (dvarrazzo) [inline tests/test_connection.py]: I don't feel we need this test, if we are not creating a reference cycle. The Connection only creates a Transaction, it doesn't hold a pointer to it, right?

File: tests/test_connection.py [hunk 18]
```diff
@@ -341,3 +349,274 @@ def cb1(n):
...
+    """
+    conn.autocommit = False
+    conn.commit()
+    conn.rollback()
+
+    with conn.transaction():
+        message = re.escape(
+            "can't change autocommit state when in Transaction context"
+        )
+        with pytest.raises(psycopg3.ProgrammingError, match=f"^{message}$"):
```
Comment (dvarrazzo) [inline tests/test_connection.py]: Don't test the messages, please. It's a chore to maintain those tests.

Comment (dvarrazzo) [reply tests/test_connection.py]: If any just test for the presence of `autocommit` in the message.

File: tests/test_connection.py [hunk 19]
```diff
@@ -341,3 +349,274 @@ def cb1(n):
...
+@pytest.mark.parametrize("autocommit", [False, True])
+def test_transaction_preserves_autocommit(conn, autocommit, exception):
+    """
+    Connection.autocommit value is False in Transaction block, but the original
+    value is always restored after the block exits, both in successful exit and
+    exception scenarios.
+    """
+    conn.autocommit = autocommit
+    with _transaction(conn, exception):
+        assert conn.autocommit is False
```
Comment (dvarrazzo) [inline tests/test_connection.py]: I really don't like this behaviour. This is an user choice, it shouldn't be messed with

File: tests/test_connection.py [hunk 20]
```diff
@@ -341,3 +349,274 @@ def cb1(n):
...
+    assert conn.autocommit is autocommit
+
+
+@pytest.mark.parametrize("exception", [False, True])
```
Comment (dvarrazzo) [inline tests/test_connection.py]: I am not sure this parametrisation and the _transaction helper is doing any good to the readability of the tests. I think these should be two different tests, really.

File: tests/test_connection.py [hunk 21]
```diff
@@ -341,3 +349,274 @@ def cb1(n):
...
+        insert_row(conn, "outer-after")
+    assert_rows(conn, {"outer-before", "outer-after"})
+    assert_rows(svcconn, {"outer-before", "outer-after"})
+
+
+def insert_row(conn, value):
+    conn.cursor().execute("INSERT INTO temp_table VALUES (%s)", (value,))
+
+
+def assert_rows(conn, expected, still_in_transaction=False):
```
Comment (dvarrazzo) [inline tests/test_connection.py]: I would leave `still_in_transaction` as a separate assert

File: tests/test_connection.py (hunk 14 above)
Comment (dvarrazzo) [reply tests/test_connection.py]: Also note that if several tests, e.g. all the tests of the same module, require the same, or few tables definitions, they can create them into a fixture.

```python
@pytest.fixture
def my_table(temp_table):
    temp_table.create("foo", "id int primary key")
    return temp_table
```

**Review by dvarrazzo — CHANGES_REQUESTED** [review]
Thank you: your tests are clearly illustrative of the feature and I think its semantic is robust! :+1:

File: psycopg3/connection.py (hunk 4 above)
Comment (asqui) [reply psycopg3/connection.py]: I like `Optional[List[str]]`. None means no active scopes; non-none is a list of active savepoints (and the empty list is a perfect fit for "no active savepoints").

Bonus 1: Will also give me an excuse to check for `None` explicitly rather than relying on truthiness (a piece of PEP 8 I personally disagree with!) :-)
Bonus 2: Allows easier auto-naming of savepoints. (I see you already spotted that bit of cumbersome logic.)

File: psycopg3/transaction.py (hunk 3 above)
Comment (asqui) [reply psycopg3/transaction.py:7]: `transaction.py` and `test_transaction.py` sounds good. Will also shorten test names as they don't all need to start with `test_transaction`. +1

File: psycopg3/connection.py (hunk 5 above)
Comment (asqui) [reply psycopg3/connection.py]: Not intentional. I just let PyCharm manage imports. Unfortunately this is an area that Black doesn't cover yet (but maybe soon? psf/black#333)

File: psycopg3/connection.py (hunk 6 above)
Comment (asqui) [reply psycopg3/connection.py]: The more general `transaction_status` is below, but this one was intended to more specifically point users to say "hey, you can't touch this when in a `with conn.transaction():` block" but maybe the more general message below is good enough?

File: psycopg3/transaction.py (hunk 12 above)
Comment (asqui) [reply psycopg3/transaction.py]: Yeah this was purely meant to avoid the curiosity of having your first savepoint named "tx_savepoint_2" in the case where an outer transaction is also started. This eye-bleed is elegantly solved through your suggestion of representing outer transaction with an empty list. +1

File: psycopg3/transaction.py (hunk 13 above)
Comment (asqui) [reply psycopg3/transaction.py]: Agree. No strong preference between the two options. Maybe stick with bytes -- I can manage some concatenation without using f-strings (also performance is bound to be massively better!)

File: tests/test_connection.py (hunk 14 above)
Comment (asqui) [reply tests/test_connection.py]: Does this objection go away if this fixture is moved to the more specific `test_transaction.py` module?

Or would you prefer I turn it into a fixture-factory, and hoist it out into `fix_db.py` so that tests in `test_connection.py` (and potentially elsewhere) can continue to use it?

If so, any preference on what to call the 'local' fixture? `temp_table_with_single_id_column_of_type_int`? :-p
(For `test_transaction.py` I'd consider making it an auto fixture, since most tests require it anyway -- this way the name becomes less of an issue as it doesn't need to be mentioned in every test signature.)

File: tests/test_connection.py (hunk 17 above)
Comment (asqui) [reply tests/test_connection.py]: Yeah this is basically a test to fail if a reference cycle is ever introduced in the future. Happy to remove if you think it has little value.

File: tests/test_connection.py (hunk 19 above)
Comment (asqui) [reply tests/test_connection.py]: Interesting thought. Maybe this is an unnecessary hangover from my previous implementation, where autocommit was the only way to control transactions so you _had_ to mess with it...

So do you think it's acceptable for autocommit to remain `False` inside a transaction block? To me this basically means "any change you make is getting committed immediately and unconditionally" which is not true inside a transaction block, but I'm also happy to say "if you're using transaction blocks, don't look at autocommit!"

File: tests/test_connection.py (hunk 21 above)
Comment (asqui) [reply tests/test_connection.py]: Agree. This was copied from the psycopg-nestedtransactions tests and I wondered why it was there. Will remove. Any test that cares should assert this separately.

File: tests/test_connection.py (hunk 19 above)
Comment (dvarrazzo) [reply tests/test_connection.py]: If you are in autocommit and people do `cur.execute("begin")` you are in a transaction too. `autocommit` only says that the connection won't start a transaction implicitly. If you start it explicitly (with a statement, or with `conn.transaction()`) it's not less auto, it's pretty manual.

File: tests/test_connection.py (hunk 17 above)
Comment (dvarrazzo) [reply tests/test_connection.py]: No, I don't think it has much value. There could be potentially cycles between any pair of objects, between connections and cursors for instance, but unless you really start referring to each other you don't really know the scenarios in which a cycle may be created. I think it's better addressed when you know when a cycle could exist.

File: psycopg3/connection.py (hunk 6 above)
Comment (dvarrazzo) [reply psycopg3/connection.py]: It may be good enough. I'm ok if there is an even better message, but I'm less ok with an extra check that can be avoided. `_savepoints` non empty implies `transaction_status` `INTRANS`: maybe you can check for it, and if we know, by the second test, that the user is in a condition of sin, we can check the `_savepoints` for a better message. The fast path, "the user is not a sinner", can be checked by just looking at `transaction_status`.

File: psycopg3/connection.py (hunk 5 above)
Comment (dvarrazzo) [reply psycopg3/connection.py]: I don't like this "import optimization" at all. Please disable it in pycharm for this project, thank you :)

File: psycopg3/connection.py (hunk 4 above)
Comment (dvarrazzo) [reply psycopg3/connection.py]: I am very ok for explicitly checking for None. `is None` also happens to be a particularly fast operation, as it doesn't drop into the whole rich comparison machinery.

File: psycopg3/transaction.py (hunk 9 above)
Comment (asqui) [reply psycopg3/transaction.py:20]: At the moment yes, but I was planning to implement the behaviour that if you specify savepoint name in outer Transaction() scope it will begin a transaction _and_ save a savepoint with your name, in which case there are two cases of "outer transaction" to differentiate.

This seems better than the other two options:
1. Ignore the savepoint name provided -- if a user is providing a name, we should respect it.
2. Raise exception saying "can't name outer transaction" -- this breaks composability; if I write a library that uses `with Transaction()` on a conn passed in, I shouldn't make restrictions on the state of that conn. Doing so brings us back to the past inconveniences of "you must pass me a connection with autocommit off for this code to work correctly" with the equivalent of "you must pass me a connection with an outer transaction already running for this code to be able to use `with Transaction(name=...)` correctly"!

So, TL;DR; true for now, but both will be needed soon :-)

Comment (dvarrazzo) [reply psycopg3/transaction.py:20]: If you have a reason to leave it this way, that's cool for me: you know the path ahead better than myself on this feature :slightly_smiling_face:

File: tests/test_connection.py (hunk 14 above)
Comment (dvarrazzo) [reply tests/test_connection.py]: Yes: the objection goes away if the tests go into a `test_transaction.py` table :) It only re-proposes itself in case you start needing more than one different table into the same module.

We can introduce the fixture-factory at a later time. I kinda liked `temp_table` for the fixture name but it's also kinda wrong (that is not a table, and is not a temp table either) so I'd probably call the fixture `tables`.

However don't implement it in this MR unless you find yourself wanting to use it.

Comment (asqui) [reply tests/test_connection.py]: How about this implementation: https://github.com/asqui/psycopg3/commit/f2e85b9914c13a5ee821cc4cfdb2e84b65c693cc
(This is a commit on a separate branch to show a cleaner diff of net changes to the original `test_connection.py`)

**Comment by asqui** [comment]
I've pushed a few commits which I believe address everything in this PR.
Let me know if you agree and I'll get on with implementing the rest of the features, which I've listed here: https://github.com/asqui/psycopg3/projects/1
